### ISSUE TITLE
[Discussion] Gradient flow

### ISSUE THREAD

**Issue by xoiga123** [datvuthanh/HybridNets#34]
Back when we were toying with mosaic, we removed the segmentation head completely from the model and dataloader. Now that we try to add mosaic augmentation officially, we have to make a decision of not using it for segmentation training. 

hybridnets/dataset.py
```python
if self.use_mosaic:
    # honestly, mosaic is not for road and lane segmentation anyway
    # you cant expect road and lane to be split up in 4 separate corners in an image, do you?
    # only use mosaic with freeze_seg :)
    img, labels, seg_label, lane_label, (h0, w0), (h, w), path = self.load_mosaic(idx)
```

Only images and object annotations are mosaic, while segmentation annotations are kept intact, which produces incorrect segmentation loss but that doesn't matter because we froze segmentation head anyway, thinking that `requires_grad=False` makes the segmentation head disappear from backprop graph. But that is [wrong](https://discuss.pytorch.org/t/will-freezing-an-intermediate-block-influence-the-gradient-flow/88859), the backbone is still affected by segmentation loss.

Check this [colab](https://colab.research.google.com/gist/xoiga123/880d3765ea40a1324a9a410ac38d345e/gradient-flow.ipynb) for interactive stuffs.

So we've been planning to just straight ahead set the losses to 0 when you --freeze_head like this:
```python
cls_loss, reg_loss, seg_loss, regression, classification, anchors, segmentation = model(imgs, annot, seg_annot, obj_list=params.obj_list)
cls_loss = cls_loss.mean() if not opt.freeze_det else 0
reg_loss = reg_loss.mean() if not opt.freeze_det else 0
seg_loss = seg_loss.mean() if not opt.freeze_seg else 0
```

Is this approach too naive? Are there any recommendation regarding this matter? Or should we also mosaic the segmentation labels?

**Comment by xoiga123** [comment]
Btw I figured it out. [Check this colab.](https://colab.research.google.com/gist/xoiga123/ac82600a8a31a92767a0eea415640a71/gradient-flow.ipynb)

1. If you freeze the last layers, backprop still flows back to the earlier layers like usual. The frozen layers just don't get the new grad -> the backbone can still learn even if you freeze all heads.
2. If segmentation head is frozen for example. the backbone still get gradient from its losses, so `freeze_seg` right now doesn't actually prevent segmentation head from affecting backbone -> must set the according losses to zero if you want to freeze properly (as if segmentation head doesn't exist). I assume this is a must for multi-task networks with different heads.
